Pods/EPContactsPicker.swift

```diff
@@ -255,160 +260,161 @@ open class EPContactsPicker: UITableViewController, UISearchResultsUpdating, UIS
-            } else {
-                predicate = CNContact.predicateForContactsInContainer(withIdentifier: contactsStore!.defaultContainerIdentifier())
-            }
-            
-            let store = CNContactStore()
-            do {
-                filteredContacts = try store.unifiedContacts(matching: predicate,
-                    keysToFetch: allowedContactKeys())
-                //print("\(filteredContacts.count) count")
-                
-                self.tableView.reloadData()
-                
-            }
-            catch {
-                print("Error!")
-            }
+  override open func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+    return 60.0
+  }
+
+  override open func tableView(_ tableView: UITableView, sectionForSectionIndexTitle title: String, at index: Int) -> Int {
+    if isSearchActive { return 0 }
+    if title.count > 1 { return 0 }
+    tableView.scrollToRow(at: IndexPath(row: 0, section: index), at: UITableView.ScrollPosition.top , animated: false)
+    return sortedContactKeys.firstIndex(of: title)!
+  }
+
+  override  open func sectionIndexTitles(for tableView: UITableView) -> [String]? {
+    if isSearchActive { return nil }
+    return sortedContactKeys.filter { $0 != externalSourceDelegate?.epExternalContactsSectionTitle() }
+  }
+
+  override open func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+    if isSearchActive { return nil }
+    return sortedContactKeys[section]
+  }
+
+  // MARK: - Button Actions
+
+  @objc func onTouchCancelButton() {
+    dismiss(animated: true, completion: {
+      self.contactDelegate?.epContactPicker(self, didCancel: NSError(domain: "EPContactPickerErrorDomain", code: 2, userInfo: [ NSLocalizedDescriptionKey: "User Canceled Selection"]))
+    })
+  }
+
+  @objc func onTouchDoneButton() {
+    dismiss(animated: true, completion: {
+      self.contactDelegate?.epContactPicker(self, didSelectMultipleContacts: self.selectedContacts)
+    })
+  }
+
+  // MARK: - Search Actions
+
+  open func updateSearchResults(for searchController: UISearchController)
+  {
+    if let searchText = resultSearchController.searchBar.text , searchController.isActive {
+
+      let predicate: NSPredicate
+      if searchText.count > 0 {
+        predicate = CNContact.predicateForContacts(matchingName: searchText)
+      } else {
+        predicate = CNContact.predicateForContactsInContainer(withIdentifier: contactsStore!.defaultContainerIdentifier())
+      }
+
+      let store = CNContactStore()
+      do {
+        filteredContacts = try store.unifiedContacts(matching: predicate,
+                                                     keysToFetch: allowedContactKeys())
+        //print("\(filteredContacts.count) count")
+        if let contacts = self.contactDelegate?.epContactPicker(self, filter: filteredContacts) {
+          filteredContacts = contacts
         }
```

Comment:
```suggestion
            filteredContacts = try store.unifiedContacts(matching: predicate,
                                                     keysToFetch: allowedContactKeys())
            let delegateFilteredContacts = filteredContacts.filter { [weak self] contact in
                self?.contactDelegate?.epContactPicker(self, shouldShow: contact)
            }
        }
```

@aly-simplehabit  I don't think we need a second delegate.  We could add a `context` or `mode` parameter to the shouldShow method which we could use to distinguish the `search` and `list` cases if needed.

Either way, we should have a consistent method where we either filter one contact at a time or filter the whole array.

Reply:
@cnnrdltn Awesome idea, i don't know why i created new delegate, the only one is enough.